src/libs/Network/SequentialQueue.js

```diff
@@ -26,18 +28,38 @@ let currentRequest = null;
 function process() {
     const persistedRequests = PersistedRequests.getAll();
 
-    // This sanity check is also a recursion exit point
-    if (NetworkStore.isOffline() || _.isEmpty(persistedRequests)) {
+    // If we have no persisted requests or we are offline we don't want to make any requests so we return early
+    if (_.isEmpty(persistedRequests) || NetworkStore.isOffline()) {
         return Promise.resolve();
     }
 
-    const task = _.reduce(persistedRequests, (previousRequest, request) => previousRequest.then(() => {
-        currentRequest = Request.processWithMiddleware(request, true);
-        return currentRequest;
-    }), Promise.resolve());
-
-    // Do a recursive call in case the queue is not empty after processing the current batch
-    return task.then(process);
+    // Get the first request in the queue and process it
+    // Make a copy of the first request in the queue and process it. We make a copy so that the middlewares can't modify the persisted request which could cause problems with retries.
```

Comment:
This comment is a bit cryptic. It feels like there is more specific explanation to share. Which part of the request gets modified? Where does it happen? And what specific problem does it cause for the retries?

Reply:
> Which part of the request gets modified? Where does it happen? And what specific problem does it cause for the retries?

There actually isn't an instance of this happening right now from what I an see. It's more of a safeguard so that if someone modifies the request within the middleware, the original won't be modified and when we retry the request. Copying the request ensure that we will always be able to retry the original, unaltered request.

Reply:
I think we should #donothing and not clone the request proactively. If someone mutates it they will introduce a bug and we can fix it then (e.g. by not directly modifying the request).

Reply:
[`Object.freeze()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze) might be another way to prevent that from happening. But I don't think I've ever actually had to use it for anything.

Reply:
That sounds reasonable. I'll remove the copying.